[16.0][FIX] auditlog: Fix Missing Log Lines When Setting Empty Field Values

BT-pgaiser opened the pull request:
Fixes the comparison with old values when setting fields to empty values.

The old values for the comparison when using **fast log** mode were initialized with `False`. This causes the problem that when a field is set to an emty value by the user, auditlog does not create a log line because internal checks compare `False` to `False` and do not consider logging the change. Full log mode is not affected by this as there everything is logged by default. However, even in fast log mode, some companies have regulatory requirements which force them to log user actions that set empty values to fields, so that should be logged properly.

BT-mlopez reviewed (approved):
LGTM, unit test look quite complete.

BT-ojossen reviewed (approved):
LGTM, but you should add yourself to the contributors

BT-pgaiser commented:
@amh-mw @StefanRijnhart 
Could someone please take a look at this and potentially merge? I don't know who's maintaining the auditlog module.

StefanRijnhart commented on auditlog/__manifest__.py:
```diff
@@ -3,7 +3,7 @@
 {
     "name": "Audit Log",
-    "version": "16.0.3.0.0",
+    "version": "16.0.3.0.1",
```
Please don´t update the version manually. Ocabot will do this during merge.

BT-pgaiser replied:
Got it. Reverted the version bump.

StefanRijnhart commented on auditlog/tests/test_auditlog.py:
```diff
@@ -321,6 +321,55 @@ def tearDown(self):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.test_model = (
```
Do you need to setup a whole model? Can't you use a res.partner field like 'website'?

BT-pgaiser replied:
I tried that but the other test cases seem to interfere with that, which results in a flaky test. I think having a separate model makes that test case more isolated and stable.

StefanRijnhart replied:
Indeed, flakiness of the tests is addressed in https://github.com/OCA/server-tools/pull/3364. I had to picked it in https://github.com/OCA/server-tools/pull/3371 to prevent tests from other modules from failing.

StefanRijnhart replied:
Would it help to move the test (based on res.partner) into its own test class?

StefanRijnhart replied:
Note that https://github.com/OCA/server-tools/pull/3364 is now merged, so maybe that helps with your test. Be sure to inherit from `AuditLogRuleCommon` and create your rule with `self.create_rule`.

BT-pgaiser replied:
Thanks. I rebased and it is working now.

StefanRijnhart replied:
Great!

StefanRijnhart reviewed (commented):
Thanks! I see the issue. But doing a read is undermining the 'fast' aspect of the 'fast' type rule. Would it work to force any falsy value in the written `vals` to show up in the log lines (conditionally overriding the result of DictDiffer)?

BT-pgaiser commented:
Thanks for the prompt response. Indeed, if we change the initial old value from `False` to `None` [here](https://github.com/OCA/server-tools/blob/5d3056492f486d7caec01d262bcf0f59c5dfa21a/auditlog/models/rule.py#L466), then the `DictDiffer` object will track the change and a log line is created. However, that log line will still show a change from "empty" to "empty" value, which is incorrect. As I have mentioned, some companies have regulatory requirements that oblige them to log these changes too.
Regarding the performance, yes, doing a round-trip to the DB has an additional cost but in many cases the field values are first read before they are written, which results in a high likelihood that on the side of the DB those values are still in cache so it won't hit the disk. Is there another way to obtain the old field values?

StefanRijnhart commented:
Isn't it the case that one never gets the old values with 'fast' logging? So having both an empty value for old and new is what is to be expected.

BT-pgaiser commented:
That seems to be the case, yeah. So this is an intentional performance optimisation? How about we add an option to the auditlog rules (applicable only for the fast mode) to let the user decide whether old values should be logged?

StefanRijnhart commented:
@BT-pgaiser Isn't that what 'full' logging is for?

BT-pgaiser commented:
If that's the only difference between "fast" and "full" log mode, then yes, simply activating the "full" mode solves the issue. I was thinking about having something in between those two modes, but maybe that does not make a lot of sense.
Can you confirm? If yes, I'd keep the fix for the missing log lines when logging empty values but remove the read of old values.

StefanRijnhart commented:
@BT-pgaiser Yes please! Keep logging the empty values when specified explicitely in the write vals (in the case of write 'fast'), but don't read old values to compare.

StefanRijnhart commented on auditlog/tests/test_auditlog.py:
```diff
@@ -322,7 +322,34 @@ def tearDown(self):
 
 
-class TestAuditlogFast(AuditLogRuleCommon, AuditlogCommon):
+class TestAuditlogFast(AuditLogRuleCommon):
```
Better add your own test because we want to keep AuditLogCommon on this one, so that it runs the standard test methods from that case.

BT-pgaiser replied:
Done.

StefanRijnhart reviewed (commented):
Can you squash commits afterwards?

BT-pgaiser committed:
[FIX] auditlog: Change comparison with old field values.

Fix the comparison with old values when setting fields to empty values.
The old values for the comparison when using fast log mode were initialized
with False. This causes the problem that when a field is set to an emty value
by the user, auditlog does not create a log line because internal checks
compare False to False and do not consider logging the change. Full log mode
is not affected by this as there everything is logged by default. However,
even in fast log mode, some companies have regulatory requirements which
force them to log user actions that set empty values to fields.

BT-pgaiser commented:
Commits squashed and test case refactored.

StefanRijnhart reviewed (approved):
Thanks, looking good now!

BT-pgaiser commented:
@StefanRijnhart Can we please merge this? Thanks.

StefanRijnhart commented:
@BT-pgaiser I would like to see another review from a tools maintainer or previous contributor to this module. The other reviews on this PR so far are clearly 'team service' because I don't see any other reviews in the OCA from them except for one or two other 'BT' PRs.

BT-pgaiser commented:
@StefanRijnhart Fair enough.
@amh-mw @mymage @tarteo @hbrunn Would one of you be so kind :)

hbrunn commented on auditlog/tests/test_auditlog.py:
```diff
@@ -343,6 +343,54 @@ def tearDown(self):
+        cls.auditlog_rule.subscribe()
+
+    def test_when_removing_field_value_a_log_line_is_created(self):
+        self.test_record.write({"phone": False})
```
can you do the same test with writing `None` here? Both are valid input, and I suspect you're just replacing when the misbehavior happens.

chances are you need some sentinel object like

```python
class FalsyValue():
    def __bool__(self):
        return False
```

to initialize the dict above with

BT-pgaiser replied:
I've added a test for writing `None` values. However, AFAIK, Odoo always represents missing/empty values as `False` and this is consistent with how the web client behaves. Anyhow, I've never seen `None` being used for that in Odoo. Surely, if I write a separate program that directly communicates with the Odoo backend, then I can send it `None` values, but I'd say that's _unspecified behaviour_ so it does not need to be supported. Anyways, the introduced sentinel object will handle both cases now.

hbrunn commented on auditlog/models/rule.py:
```diff
@@ -463,7 +473,7 @@ def write_fast(self, vals, **kwargs):
             # of the data in the database
             vals2 = dict(vals)
-            old_vals2 = dict.fromkeys(list(vals2.keys()), None)
+            old_vals2 = dict.fromkeys(list(vals2.keys()), _Sentinel())
```
wait, for better performance, please create an instance of this above and recycle this object, no need to keep creating it

BT-pgaiser replied:
Done.

BT-pgaiser committed:
[FIX] auditlog: Change comparison with old field values.

Introduce sentinel object as a placeholder for absent values
when comparing diff.
Add test for setting None values.

BT-pgaiser commented:
@hbrunn Sorry for the delay, I was OoO last week.

hbrunn reviewed (approved):
thank you!

/ocabot merge patch
